File: tests/lang-test/cpp/models/product.cpp
// ripex-lang-test: C++ Product impl — template definition.
#include "product.hpp"

template <typename T>
Product<T>::Product(int id, std::string name, double price, T category)
    : id_(id), name_(std::move(name)), price_(price), category_(category) {}

template <typename T>
double Product<T>::calculate_tax(double rate) const {
    return price_ * rate;
}

template class Product<Category>;

File: tests/lang-test/rust/src/models/product.rs
//! ripex-lang-test: Rust Product — enum, generic impl, const.
#[derive(Debug)]
pub enum Category {
    Electronics,
    Clothing,
}

pub struct Product<T> {
    pub id: u32,
    pub name: String,
    pub price: f64,
    pub category: T,
}

impl<T> Product<T> {
    pub const TAX_RATE: f64 = 0.1;

    pub fn calculate_tax(&self) -> f64 {
        self.price * Self::TAX_RATE
    }
}

pub fn make_widget() -> Product<Category> {
    Product {
        id: 1,
        name: "Widget".to_string(),
        price: 19.99,
        category: Category::Electronics,
    }
}
